File: src/topic02_motion/my_attempt/motion_my_attempt.py
#!/usr/bin/env python3

import rospy
import math
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose

def callback_pose(pose_msg):
    global x,y, theta
    x = pose_msg.x
    y = pose_msg.y
    theta = pose_msg.theta

def move(topic, distance, speed, is_forward=True):
    velocity_message = Twist()
    velocity_publisher = rospy.Publisher(topic, Twist, queue_size=10)
    
    if is_forward:
        velocity_message.linear.x = speed
    else:
        velocity_message.linear.x = -speed
    
    global x,y
    x0 = x
    y0 = y
    distance_travelled = 0
    rate = rospy.Rate(10)
    
    while(distance_travelled < distance):
        velocity_publisher.publish(velocity_message)
        rate.sleep()
        distance_travelled = math.sqrt((x-x0)**2 + (y-y0)**2)
        rospy.loginfo("Target:%f, Travelled:%f"%(distance,distance_travelled))
    
    velocity_message.linear.x = 0
    velocity_publisher.publish(velocity_message)
    rospy.loginfo("Reached")
    return "Reached"

def rotate(topic, angle_rad, angular_speed_rad, anti_clockwise=True):
    angular_velocity_message = Twist()
    angular_velocity_publisher = rospy.Publisher(topic, Twist, queue_size=10)
    
    if anti_clockwise:
        angular_velocity_message.angular.z = angular_speed_rad
    else:
        angular_velocity_message.angular.z = -angular_speed_rad
    
    global theta
    target_angle = (theta + angle_rad)%(2*math.pi)
    if(target_angle > math.pi):
        target_angle -= 2*math.pi
    elif(target_angle < -math.pi):
        target_angle += 2*math.pi

    rate = rospy.Rate(10)
    
    while(abs(theta-target_angle)>0.01):
        angular_velocity_publisher.publish(angular_velocity_message)
        rospy.loginfo("Relative:%s, target:%s, current:%s"%(angle_rad, target_angle,theta))
        
        rate.sleep()
        
    angular_velocity_message.angular.z = 0
    angular_velocity_publisher.publish(angular_velocity_message)
    rospy.loginfo("Rotation Finished")
    return "Finished"
    
def go_to(topic, x_goal, y_goal):
    global x,y,theta
    angle_of_rotation = math.atan2(y_goal-y, x_goal-x) - theta

    angular_speed = 0.2
    if(angle_of_rotation > 0): 
        anti_clockwise = True
    else:
        anti_clockwise = False

    rotate(topic, angle_of_rotation, angular_speed, anti_clockwise)
    
    distance = math.sqrt((x_goal-x)**2 + (y_goal-y)**2)
    
    speed = 1
    move(topic, distance, speed, True)
    
    rospy.loginfo("Destination Reached")
    return "Reached"

if __name__ == "__main__":
    rospy.init_node("motion_my_attempt")
    
    pose_topic = "/turtle1/pose"
    pose_subscriber = rospy.Subscriber(pose_topic, Pose, callback_pose)
    
    rospy.sleep(0.1)
    
    velocity_topic = "/turtle1/cmd_vel"
    # rotate(velocity_topic, 1, 1)
    # move(velocity_topic, 3, 1)
    go_to(velocity_topic, 0,0)

File: src/topic02_motion/my_attempt/motion_my_attempt.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <turtlesim/Pose.h>
#include <cmath>

ros::Subscriber pose_subscriber;
ros::Publisher velocity_publisher;

turtlesim::Pose current_pose;
geometry_msgs::Twist vel_msg;

const float Kp_linear = 1;
const float Kp_angular = 1;
const double PI = 3.14159265359;

void poseCallback(turtlesim::Pose pose_msg);
void move(double distance, bool is_forward);
void rotate(double angle_rad, bool ccw);
void go_to(double x, double y);

int main (int argc, char** argv)
{
    ros::init(argc, argv, "motion_my_attempt");
    ros::NodeHandle nh;
    
    pose_subscriber = nh.subscribe("/turtle1/pose", 10, poseCallback);
    velocity_publisher = nh.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);

    ros::Duration(1).sleep();
    ros::spinOnce();

    ROS_INFO("Node motion_my_attempt up running");

    move(4, true);
    rotate(3.14, true);
    move(4, true);
    rotate(1, true);
    move(1, true);
    rotate(-1, false);
    move(1, true);

    go_to(5.544445, 5.544445);
}

void poseCallback(turtlesim::Pose pose_msg) {
    current_pose.x = pose_msg.x;
    current_pose.y = pose_msg.y;
    current_pose.theta = pose_msg.theta;
}

void move(double target_distance, bool is_forward) {
    float x0 = current_pose.x;
    float y0 = current_pose.y;
    double distance_travelled = 0.0;
    double gap = target_distance - distance_travelled;
    int forward_flag = 1;
    ros::Rate rate(10);

    if(!is_forward) forward_flag = -1;

    while(true)
    {
        vel_msg.linear.x = Kp_linear * gap * forward_flag;
        velocity_publisher.publish(vel_msg);

        // ROS_INFO("x0:%f, x:%f, y0:%f, y:%f",x0, current_pose.x, y0, current_pose.y);
        distance_travelled = sqrt(pow(current_pose.x-x0,2) + pow(current_pose.y-y0,2));
        gap = target_distance - distance_travelled;
        
        ROS_INFO("Travelled:%f, gap:%f", distance_travelled, gap);
        if(abs(gap) < 0.05) break;

        ros::spinOnce();
        rate.sleep();
    }

    vel_msg.linear.x = 0;
    velocity_publisher.publish(vel_msg);
    ROS_INFO("Reached");
}

void rotate(double angle_rad, bool ccw) {
    double target_angle = current_pose.theta + angle_rad;
    ROS_INFO("Target_angle:%f",target_angle);
    int ccw_flag = 1;
    ros::Rate rate(100);

    if(!ccw) ccw_flag = -1;

    if(target_angle > PI) target_angle = target_angle - (2*PI);
    else if(target_angle < -PI) target_angle = 2*PI + target_angle;

    double gap = target_angle - current_pose.theta;
    while(true)
    {
        ROS_INFO("Gap:%f, current:%f",gap,current_pose.theta);
        vel_msg.angular.z = Kp_angular * abs(gap) * ccw_flag;
        velocity_publisher.publish(vel_msg);

        gap = target_angle - current_pose.theta;
        if(abs(gap) < 0.01) break;

        ros::spinOnce();
        rate.sleep();
    }

    vel_msg.angular.z = 0;
    velocity_publisher.publish(vel_msg);
    ROS_INFO("Finished");
}

void go_to(double x, double y) {
    double delta_y = y-current_pose.y;
    double delta_x = x-current_pose.x;
    bool ccw;

    double delta_theta = atan2(delta_y,delta_x) - current_pose.theta;
    if(delta_theta>=0 && delta_theta<=PI) ccw=true;
    else ccw=false;
    rotate(delta_theta, ccw);

    double distance = sqrt(pow(delta_y,2)+pow(delta_x,2));
    move(distance, true);

    ROS_INFO("Reached Destination");
}